pydis_site/apps/content/models/tag.py

```diff
@@ -0,0 +1,75 @@
+import collections.abc
+import json
+
+from django.db import models
+
+
+class Commit(models.Model):
+    """A git commit."""
+
+    URL_BASE = "https://github.com/python-discord/bot/commit/"
+
+    sha = models.CharField(
+        help_text="The SHA hash of this commit.",
+        primary_key=True,
+        max_length=40,
+    )
+    message = models.TextField(help_text="The commit message.")
+    date = models.DateTimeField(help_text="The date and time the commit was created.")
+    author = models.TextField(help_text="The person(s) who created the commit.")
+
+    @property
+    def url(self) -> str:
+        """The URL to the commit on GitHub."""
+        return self.URL_BASE + self.sha
+
+    def lines(self) -> collections.abc.Iterable[str]:
+        """Return each line in the commit message."""
+        for line in self.message.split("\n"):
+            yield line
+
+    def format_users(self) -> collections.abc.Iterable[str]:
+        """Return a nice representation of the user(s)' name and email."""
+        for author in json.loads(self.author):
+            yield f"{author['name']} <{author['email']}>"
+
+
+class Tag(models.Model):
+    """A tag from the python-discord server."""
+
+    URL_BASE = "https://github.com/python-discord/bot/tree/main/bot/resources/tags"
+
+    last_updated = models.DateTimeField(
+        help_text="The date and time this data was last fetched.",
+        auto_now=True,
+    )
+    sha = models.CharField(
+        help_text="The tag's hash, as calculated by GitHub.",
+        max_length=40,
+    )
+    last_commit = models.ForeignKey(
+        Commit,
+        help_text="The commit this file was last touched in.",
+        null=True,
```

Comment:
The initial value can be null, since the data is only populated when requesting a single tag page (to avoid making 100 requests at once). Once the tag is accessed for the first time (since the last update), it's commit data is saved for future access.

Reply:
Relevant code in the get_tag method:
https://github.com/python-discord/site/blob/92a42694b6ad1a29e5a21e0b3e57639528837113/pydis_site/apps/content/utils.py#L245-L248